### ISSUE TITLE
[Feature]: Add an Iris dataset provider (and a tabular RawDataset -> Dataset bridge)

### ISSUE THREAD

**Issue by michalharakal** [sk-ai-net/skainet#1044]
> 👋 **New to SKaiNET? This one is for you.**
>
> This issue is deliberately written to be a *first* contribution. It touches one module, needs
> no maths or GPU knowledge, and there is an existing file in the codebase you can copy the
> structure from almost line by line. If anything below is unclear, that is a bug in the issue,
> not in you — please just ask in the comments and we will fix the wording.
>
> **Estimated effort:** half a day for Part 1. **Prerequisites:** Kotlin basics. That is it.

---

# 🧠 D: DOCUMENT — Problem & Opportunity

**Summary (1–2 sentences):**
> SKaiNET ships ready-made dataset providers for MNIST, Fashion-MNIST and CIFAR-10, but nothing
> for small tabular datasets. Anyone writing a classification example has to hand-write CSV
> parsing, one-hot encoding and tensor batching before they can touch the interesting part.

## What is missing

Loading MNIST in SKaiNET today is one line:

```kotlin
val train = MNIST.loadTrain()   // suspend; downloads, caches, returns a Dataset
```

There is no equivalent for Iris — or for *any* tabular dataset. The pieces exist but do not meet:

- ✅ `skainet-data-source` parses CSV into a `RawDataset` (a `List<RawDataRow>`, where each row is
  a `Map<String, String>`, plus a `DataSchema`).
- ✅ `skainet-data-api` defines `Dataset<X, Y>`, which gives you `split`, `shuffle`, `batches`,
  `epochs`, `filter` and `mapX`/`mapY` for free.
- ❌ **Nothing converts the first into the second.** `RawDataset` holds strings; `Dataset` must
  emit `DataBatch` tensors. That conversion is left to every user.

## Why this matters now

PR [#34](https://github.com/SKaiNET-developers/SKaiNET-examples/pull/34) from [@EdinEd](https://github.com/EdinEd) added a lovely Iris
multiclass example — and roughly half of its ~200 lines are data plumbing:

| Hand-written in the example | Already in SKaiNET |
|---|---|
| `loadKaggleCsv()` | `DataFormat.CSV` parser in `skainet-data-source` |
| `test_train_split()` | `Dataset.split(ratio, seed, stratified)` |
| `labelToOneHot()` | — (genuinely missing) |
| manual per-sample tensor construction | `Dataset.batches()` |

The same boilerplate is being reproduced in the accompanying teaching notebook, where it is
actively in the way — the notebook has to spend a whole section on `Dataset` plumbing before it
can talk about classification.

## Who benefits

- **Newcomers** — the first Kotlin ML example someone reads shouldn't open with a CSV parser.
- **Example and notebook authors** — one line instead of forty, in every tabular example.
- **Everyone with a CSV** — Part 2 below makes *any* tabular file loadable, not just Iris.

---

# 🔍 A: ASSESS — Feasibility & Impact

### ✔️ Feasibility

**High.** This is additive: no existing API changes, nothing to deprecate, no numerical work.
`MNISTDataset` in
`skainet-data/skainet-data-simple/src/commonMain/kotlin/sk/ainet/data/mnist/MNISTData.kt`
is a direct template — Iris is strictly simpler (no gzip, no IDX binary format, no image layout).

Iris is ~4 KB and public domain, so unlike MNIST it can simply be **embedded as a resource**.
No download, no cache directory, no network in tests. That removes most of what makes
`MNISTLoader` complicated, which is exactly why this makes a good first issue.

### ✔️ Expected Impact

- Every tabular example and notebook loses its boilerplate section.
- Part 2 unlocks CSV/TSV/JSON-Lines → `Dataset` generally, which is a frequently requested path.
- Gives SKaiNET the "toy dataset" on-ramp that scikit-learn's `load_iris()` provides.

### ✔️ Risks / Constraints

| Risk | Mitigation |
|---|---|
| Class-index ordering must be stable | Fix alphabetical order in a documented constant; assert in tests |
| `Dataset` `Y` type choice affects stratification | Use `Int` class index — see the trap in Research below |
| Feature column order must be stable | Define explicitly, never rely on `Map` iteration order |
| Embedding data in a jar | 4 KB, public domain — negligible |

### ✔️ Dependencies

- **Part 1:** `skainet-data-api` (for `Dataset`/`DataBatch`) and `skainet-lang-core` (tensors).
  Lands in `skainet-data-simple` beside the MNIST/CIFAR providers.
- **Part 2:** additionally `skainet-data-source` (for `RawDataset`).
- No third-party libraries.

---

# 📚 R: RESEARCH — What Must Be Understood First?

### Reading list (all in this repo — start here)

- [ ] `skainet-data/skainet-data-api/src/commonMain/kotlin/sk/ainet/data/Dataset.kt` — **the one
      file to actually read**. Note which members are `abstract` (you must implement them) versus
      `open`/concrete (you get them free).
- [ ] `skainet-data/skainet-data-api/src/commonMain/kotlin/sk/ainet/data/DataBatch.kt` — what you
      must produce.
- [ ] `skainet-data/skainet-data-simple/.../mnist/MNISTData.kt` — the pattern to copy.
- [ ] `skainet-data/skainet-data-simple/.../mnist/MNIST.kt` — the tiny public entry-point object.
- [ ] `skainet-data/skainet-data-source/.../DataFormatParser.kt` — `RawDataset` shape (Part 2 only).

### ⚠️ Three traps worth knowing before you start

**1. You must override `createIndexedDataBatch`, not just `createDataBatch`.**
`split()` and `shuffle()` return an *index view* over your dataset rather than copying data. When
that view asks for a batch, it passes **non-contiguous** indices. The base-class default only
handles contiguous ranges and throws otherwise. Override both and shuffled/split/filtered views
all work; override only the first and `split(...)` then `batches(...)` fails at runtime.

**2. `Y` should be `Int` (the class index), not the one-hot `FloatArray`.**
Stratified splitting buckets samples in a `LinkedHashMap` keyed by `getY(idx)`. `FloatArray`
compares by *identity*, so a one-hot label would put every sample in its own bucket and silently
defeat stratification — no error, just a split that isn't stratified. Keep `getY` returning the
class index; do the one-hot conversion inside batch construction.

**3. Cross-entropy expects logits, not probabilities.**
Not your code, but relevant to the tests: `CategoricalCrossEntropyLoss` applies log-softmax
internally. If you write a smoke test that trains a small model, do **not** put a softmax on the
output layer.

### Prior art

- [ ] scikit-learn `load_iris()` — returns data/target/feature_names/target_names. Worth mirroring
      the *metadata* idea: expose `featureNames` and `classNames`, not just numbers.
- [ ] PyTorch `torchvision.datasets` — `Dataset` + transform separation, which SKaiNET already follows.

### Open questions (comment if you have an opinion)

> - Should `Iris.load()` be `suspend`? MNIST's is, because it downloads. Embedded Iris needn't be —
>   but a non-suspend outlier breaks call-site symmetry. **Current lean:** make it `suspend` for
>   consistency, since callers are usually in `runBlocking` anyway.
> - Should the provider standardise features, or return raw centimetres? **Current lean:** raw.
>   Scaling statistics must be fitted on the training split only, so the provider is the wrong
>   place — it doesn't know where you'll split.
> - Does Part 2's adapter belong in `skainet-data-source` or a new `skainet-data-tabular` module?

---

# 🛠️ C: CODE — Implementation Plan

**Part 1 and Part 2 are independent.** Part 1 alone is a complete, mergeable contribution — please
do not feel obliged to take both. If you only want Part 1, say so in a comment and someone else
can pick up Part 2.

## Part 1 — `Iris` provider *(the good-first-issue part)*

**Location:** `skainet-data/skainet-data-simple/src/commonMain/kotlin/sk/ainet/data/iris/`

### Proposed API

```kotlin
package sk.ainet.data.iris

/** One Iris flower: four measurements in cm plus its species index. */
public data class IrisSample(
    val sepalLength: Float,
    val sepalWidth: Float,
    val petalLength: Float,
    val petalWidth: Float,
    val label: Int          // 0=setosa, 1=versicolor, 2=virginica
)

/**
 * Iris dataset (Fisher, 1936). 150 samples, 4 features, 3 balanced classes.
 *
 *   x -> Tensor<FP32, Float> [batch, 4]
 *   y -> Tensor<FP32, Float> [batch, 3]  (one-hot)
 */
public data class IrisDataset(
    val samples: List<IrisSample>,
    private val executionContext: ExecutionContext = DefaultDataExecutionContext()
) : Dataset<FloatArray, Int>() { /* ... */ }

/** Entry point, mirroring `MNIST`. */
public object Iris {
    public val featureNames: List<String> =
        listOf("sepalLength", "sepalWidth", "petalLength", "petalWidth")

    public val classNames: List<String> =
        listOf("Iris-setosa", "Iris-versicolor", "Iris-virginica")

    /** Loads the bundled copy of the dataset. */
    public suspend fun load(): IrisDataset
}
```

Usage — replacing ~40 lines with one:

```kotlin
val (train, test) = Iris.load().split(0.8, seed = 42L, stratified = true)
```

### Development tasks

- [ ] Add `iris.csv` (150 rows, public domain) as a resource in `skainet-data-simple`
- [ ] `IrisSample` data class
- [ ] `IrisDataset : Dataset<FloatArray, Int>` implementing `xSize`, `getX`, `getY`, `shuffle()`,
      `split(ratio)`, `createDataBatch(...)` **and `createIndexedDataBatch(...)`** (trap 1 above)
- [ ] `Iris` object with `load()`, `featureNames`, `classNames`
- [ ] Register the module in the docs' module table if needed

### Tests

- [ ] `load()` returns exactly 150 samples, 50 per class
- [ ] Feature values fall in the documented ranges (sepal length 4.3–7.9, petal width 0.1–2.5)
- [ ] Class indices agree with `classNames` ordering
- [ ] `dataBatch` produces tensors of shape `[n, 4]` and `[n, 3]`; each `y` row sums to 1.0
- [ ] `split(0.8, seed = 42L, stratified = true)` gives 120/30 with 40/10 per class
- [ ] **Regression test for trap 1:** `split(...)` *then* `batches(...)` must not throw
- [ ] Same seed ⇒ identical splits across runs

### Acceptance criteria

- [ ] `Iris.load()` works with no network access
- [ ] Public API has KDoc; follows the `MNIST` naming pattern
- [ ] Passes `explicit-api` and binary-compatibility-validator checks
- [ ] No regressions elsewhere

## Part 2 — tabular `RawDataset` → `Dataset` bridge *(follow-up; separate PR welcome)*

Generalises Part 1 so any CSV/TSV/JSON-Lines source becomes a `Dataset`.

```kotlin
val dataset: Dataset<FloatArray, Int> = rawDataset { from("file:///data/wine.csv") }
    .toTabularDataset {
        features("alcohol", "malicAcid", "ash")   // column order is explicit
        label("cultivar")                          // string labels -> stable indices
        oneHotLabels()                             // or labelIndices() for Int32 targets
    }
```

- [ ] Design the builder scope (`features`, `label`, `oneHotLabels`/`labelIndices`)
- [ ] Implement `TabularDataset : Dataset<FloatArray, Int>` (Part 1's class, generalised)
- [ ] Decide module placement (see Open Questions)
- [ ] Handle missing values and unparseable cells with a clear error naming row *and* column
- [ ] Tests: numeric + string labels, missing columns, malformed rows, schema mismatch
- [ ] Reimplement Part 1's `IrisDataset` on top of it, keeping the `Iris.load()` API unchanged
- [ ] Add a `how-to` page under `docs/modules/ROOT/pages/how-to/`

---

# 💬 Additional Notes

### Reference implementation

A complete working `IrisDataset` — including both `createDataBatch` overrides — already exists in
the multiclass-classification teaching notebook that accompanies this issue. It is written against
released **0.40.1** and can be lifted almost verbatim into `skainet-data-simple`. Treat it as a
starting point, not a specification.

### Why this is a good first issue

- **One module, additive.** Nothing existing changes, so review is low-risk.
- **A template to copy.** `MNISTData.kt` does the same job in a harder setting.
- **Obvious tests.** Iris has 150 rows and known-good values; assertions write themselves.
- **Visible payoff.** Your PR deletes forty lines from every tabular example that follows.

### Getting started

```bash
git clone https://github.com/SKaiNET-developers/SKaiNET.git
cd SKaiNET
./gradlew :skainet-data:skainet-data-simple:jvmTest
```

See [`CONTRIBUTING.md`](../CONTRIBUTING.md) and [`GITFLOW.adoc`](../GITFLOW.adoc) — branch off
`develop`. Please comment before starting so we can assign it to you and avoid duplicate work.

### Related

- SKaiNET-examples PR #34 — the Iris example that motivated this
- Companion notebook: *Multiclass Classification with SKaiNET*
- Existing providers: `sk.ainet.data.mnist`, `sk.ainet.data.fashionmnist`, `sk.ainet.data.cifar10`

**Comment by AjithGoveas** [comment]
Hi @michalharakal ! 👋
I'd love to take on **Part 1 (Iris dataset provider)**.
I've gone through the issue description and the `MNISTData.kt` reference. To make it work seamlessly across all KMP targets (JVM, Native, JS/Wasm) without extra resource loading infrastructure, I plan to embed the 150-row CSV as an `internal const val IRIS_CSV` string in `IrisData.kt` and implement the batch delegation for `createIndexedDataBatch` as noted.
I'll branch off `develop` (`feature/1044-iris-dataset-provider`) following the repo's guidelines.
Could you please assign this issue to me?

**Comment by michalharakal** [comment]
Thank you @AjithGoveas 
I was already palying a liitle bit on this with Kotlin Notebook, take is only as an inspration too https://github.com/SKaiNET-developers/skainet-notebook/blob/develop/notebooks/classification/MulticlassClassification_Iris.ipynb

Looking forward to you PR.

**Comment by AjithGoveas** [comment]
Hi @michalharakal , I’ve opened this PR to address #1044. Let me know if you’d like any changes.

**Comment by michalharakal** [comment]
@AjithGoveas thank you very much for you contribution

**Comment by AjithGoveas** [comment]
Thanks for the review and fix! I’ll keep the KDoc notes in mind for future contributions. ❤️
